cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowProgramRunner.java

```diff
@@ -68,15 +96,48 @@ public ProgramController run(Program program, ProgramOptions options) {
 
     WorkflowSpecification workflowSpec = appSpec.getWorkflows().get(program.getName());
     Preconditions.checkNotNull(workflowSpec, "Missing WorkflowSpecification for %s", program.getName());
-    
-    WorkflowDriver driver = new WorkflowDriver(program, options, hostname, workflowSpec, programRunnerFactory);
 
     // Controller needs to be created before starting the driver so that the state change of the driver
     // service can be fully captured by the controller.
     RunId runId = RunIds.fromString(options.getArguments().getOption(ProgramOptionConstants.RUN_ID));
+
+    AdapterDefinition adapterSpec = getAdapterSpecification(options.getArguments());
+
+    WorkflowActionSpecification newActionSpec = new DefaultWorkflowActionSpecification(workflowSpec.getClassName(),
+                                                                                       workflowSpec.getName(),
+                                                                                       workflowSpec.getDescription(),
+                                                                                       workflowSpec.getProperties());
+
+    BasicWorkflowActionContext context =
+      new BasicWorkflowActionContext(newActionSpec, program, runId, 1, options.getUserArguments(), cConf,
+                                     metricsCollectionService, datasetFramework, txClient, discoveryServiceClient,
+                                     streamWriterFactory, adapterSpec,
+                                     createPluginInstantiator(adapterSpec, program.getClassLoader()));
+
+    WorkflowDriver driver = new WorkflowDriver(program, options, hostname, workflowSpec, programRunnerFactory, context);
+
     ProgramController controller = new WorkflowProgramController(program, driver, serviceAnnouncer, runId);
     driver.start();
 
     return controller;
   }
+
+  @Nullable
+  private AdapterDefinition getAdapterSpecification(Arguments arguments) {
+    // TODO: Refactor ProgramRunner class hierarchy to have common logic moved to a common parent.
```

Comment:
`getAdapterSpecification` and `createPluginInstantiator` methods used somewhere? if not we can remove them.